Software/DumptruckUltra-MCU/logic/driving-algorithm/DrivingAlgorithm.cpp

```diff
@@ -0,0 +1,130 @@
+        if (distanceToTarget(currPose) > TARGET_PROXIMITY_THRESHOLD_METERS || (currentTarget.heading - currPose.heading) > TARGET_HEADING_THRESHOLD_RADIANS) {
+            // Find target heading (if we are already close to the target, turn to match the goal heading)
+            const float headingToTarget{(distanceToTarget(currPose) > TARGET_PROXIMITY_THRESHOLD_METERS)
+                                            ? std::atan2(currentTarget.y - currPose.y, currentTarget.x - currPose.x)
+                                            : currentTarget.heading};
+            // If something is in the way...
+            if (getFrontDistanceFunction() < DISTANCE_THRESHOLD_METERS) {
+                // Turn right
+                leftMotor.setPower(Hardware::Motors::Motor::MOTOR_MAX_SPEED_ABS);
+                rightMotor.setPower(-Hardware::Motors::Motor::MOTOR_MAX_SPEED_ABS);
+            } else {
+                // Otherwise, drive at the target
+                const auto motorPowers{deltaAngleToDrivePowers(headingToTarget - currPose.heading)};
+                leftMotor.setPower(motorPowers.leftSpeed);
+                rightMotor.setPower(motorPowers.rightSpeed);
+            }
+        } else {
+            // We are at our current target, so stop, signal complete, and wait for restart
+            stop(DrivingAlgorithmStatus::COMPLETE);
+        }
+        // Req'd to prevent constantly resetting the PWM and destroying the waveform
+        vTaskDelay(DRIVING_ALGORITHM_TASK_PERIOD_TICKS);
+    }
+}
+
+/* Different methods are here: https://www.desmos.com/calculator/odrruerlc5
+ * In short:
+ *  - If the target is to the left, slow down the left wheel and keep the right wheel at full power
+ *  - Otherwise, slow down the right wheel and keep the left wheel at full power
+ * This change happens linearly with both wheels at maximum power at:
+ *  - 0 (driving forward)
+ *  - pi/2 (turning right)
+ *  - pi/2 (turning left)
+ */
+auto DrivingAlgorithm::deltaAngleToDrivePowers(float angleDiff) -> DrivingAlgorithm::DrivingPower {
+    float leftSpeed{0};
+    float rightSpeed{0};
+    constexpr float CLIMB_RATE{4};
+    if (angleDiff < 0) {
+        leftSpeed = Hardware::Motors::Motor::MOTOR_MAX_SPEED_ABS * std::max(static_cast<float>(CLIMB_RATE / M_PI * angleDiff + 1), -1.0F);
+        rightSpeed = Hardware::Motors::Motor::MOTOR_MAX_SPEED_ABS;
+    } else {
+        leftSpeed = Hardware::Motors::Motor::MOTOR_MAX_SPEED_ABS;
+        rightSpeed = Hardware::Motors::Motor::MOTOR_MAX_SPEED_ABS * std::max(static_cast<float>(-CLIMB_RATE / M_PI * angleDiff + 1), -1.0F);
+    }
+    return {.leftSpeed = leftSpeed, .rightSpeed = rightSpeed};
+}
+
+auto DrivingAlgorithm::distanceToTarget(const DeadReckoning::Pose2D &currPosition) const -> float {
+    // Assume Euclidean (planar) distance
+    constexpr float SQUARE_POWER{2};
+    return std::sqrt(std::pow(currPosition.x - currentTarget.x, SQUARE_POWER) + std::pow(currPosition.y - currentTarget.y, SQUARE_POWER));
```

Comment:
when would this ever be different than 2? can we just simplify?

Reply:
No.  Linter requirement.  The compiler will optimize this anyways.